eos-phone-home

```diff
@@ -0,0 +1,406 @@
+
+        request_data = json.dumps(request)
+        print("Request data:", request_data)
+
+        print("Sending to server (%s)..." % endpoint)
+
+        if self._debug:
+            print("Debugging turned on so nothing sent to server!")
+            return True
+
+        success = False
+
+        try:
+            data = request_data.encode('utf-8')
+
+            req = urllib.request.Request(endpoint, data=data, method='PUT')
+            req.add_header('Content-Type', 'application/json')
+
+            response_data = None
+            with urllib.request.urlopen(req) as conn:
+                response_data = conn.read().decode('utf-8')
+
+            print('Response:', response_data)
+
+            response = json.loads(response_data)
+
+            if 'success' in response and \
+               response['success'] == True:
+                print("Server returned success message!")
+                success = True
+            else:
+                print("Server failed to process the data:", response)
+
+        except Exception as e:
+            print("Sending failed!", e, file=sys.stderr)
+
+        return success
+
+    def _need_to_activate(self):
+        return not os.path.exists(os.path.join(self.STATE_DIRECTORY,
+                                               'activated'))
+
+    def _do_activate(self):
+        if self._send_to_server(self.ACTIVATION_ENDPOINT,
+                                self.ACTIVATION_VARIABLES):
+            # We're done, mark us activated
+            with open(os.path.join(self.STATE_DIRECTORY, 'activated'), 'w'):
+                pass
+
+            print("Activated!")
+        else:
+            print("Could not send activation data to server!", file=sys.stderr)
+
+            return False
+
+        return True
+
+    def _need_to_ping(self):
+        if self._lookup_or_get_variable('live'):
+            print("Not sending ping from live system.")
+            return False
+
+        if not os.path.exists(os.path.join(self.STATE_DIRECTORY, 'count')):
+            print("Creating new count file!")
+            self._set_count(0)
+            return True
+
+        count_time = os.path.getmtime(os.path.join(self.STATE_DIRECTORY,
+                                                   'count'))
+        count_age = time.time() - count_time
+        print("Count age:", count_age)
+
+        if count_age < 0:
+            print("Clock has gone backwards? Resetting time so we will ping",
+              "again in 24 hours.")
+
+            count = self._lookup_or_get_variable('count')
+            self._set_count(count)
+
+            return False
+
+        # we aim to ping daily
+        DAY_IN_SECONDS = 60 * 60 * 24
+        return count_age >= DAY_IN_SECONDS
+
+    def _do_ping(self):
+        if self._send_to_server(self.PING_ENDPOINT, self.PING_VARIABLES):
+            # increment the counter upon successful ping only
+            count = self._lookup_or_get_variable('count')
```

Comment:
@ramcq So reading more into this makes me think that we might have some problems with the current approach:
- If we only increment pings on success of sending, that means that the count sent to server isn't representative of the total pings so the length of time that the machine has been in service cannot be known from this information. If we have a lot of machines that have unstable internet (which we will), we may see some really unrepresentative numbers here of length of time in service but server-side logic will be pretty easy.
- If we increment ping one for each day (that we were installed from) even if the sending fails, the server logic is a bit harder for buckets but we can get a more accurate time of the length of service that the machine was on so we can map accurately if users are dropping off or not.

I'd opt for the latter here or just removing the count completely and counting the number of machines that pinged us during the period of one day as the metric.